### PR TITLE
desktopfile: Change validity check

### PR DESCRIPTION
**Pull request by palinek**
Based on Qt doc:
'Note: when dealing with remote objects, it is not always possible to
determine if it exists when creating a QDBusInterface.'
The QDBusAbstractInterface::lastError() seems to be more appropriate.

### CODE REVIEW

**Comment by palinek** [comment]
...let's make the testing in PR to avoid mess in master (as it seems I was too hurry commiting to master).

**Comment by palinek** [comment]
Actualy this is also not 100%...
Example: nautilus, steps to reproduce:
- start the nautilus from e.g. lxqt-panel (mainmenu or quicklaunch)
- close nautilus window
- wait until the org.gnome.nautilus disappears from session dbus (e.g. looking in qdbusviewer)
- try to start nautilus as in step 1 -> starting will fail: `XdgDesktopFileData::startByDBus: can't call method on invalid interface: "Could not get owner of name 'org.gnome.Nautilus': no such name"`

This is the latter DBus communication:
```
method call time=1456917945.031316 sender=:1.145 -> destination=org.freedesktop.DBus serial=25 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=GetNameOwner
   string "org.gnome.Nautilus"
error time=1456917945.031355 sender=org.freedesktop.DBus -> destination=:1.145 error_name=org.freedesktop.DBus.Error.NameHasNoOwner reply_serial=25
   string "Could not get owner of name 'org.gnome.Nautilus': no such name"
method call time=1456917945.031518 sender=:1.145 -> destination=org.freedesktop.DBus serial=26 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=AddMatch
   string "type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='org.gnome.Nautilus'"
method call time=1456917945.031629 sender=:1.145 -> destination=org.freedesktop.DBus serial=27 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=RemoveMatch
   string "type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='org.gnome.Nautilus'"
```

@luis-pereira @paulolieuthier Should we check the validity of `QDBusInterface` at all? (if the check is removed, in the above scenario nautilus is spawned correctly)

**Comment by luis-pereira** [comment]
@palinek Do you have any clue on the origin of that error ? 
I'm getting a `QDBusError::InvalidMember` in the second run. That's weird. 

It deserves more investigations, In the meantime we could only ignore `QDBusError::InvalidMember`
Done here:
https://github.com/luis-pereira/libqtxdg/tree/dbus_activate https://github.com/luis-pereira/libqtxdg/commit/741111bbf409827ed47f4b727be1704fff9ddf9d

File: xdgdesktopfile.cpp [hunk 1]
```diff
@@ -462,7 +462,7 @@ bool XdgDesktopFileData::startByDBus(const QStringList& urls) const
         return false;
     }
     QDBusInterface app(f.completeBaseName(), path, QLatin1String("org.freedesktop.Application"));
-    if (!app.isValid())
+    if (app.lastError().isValid())
```
Comment (paulolieuthier) [inline xdgdesktopfile.cpp]: I'd write `if (!app.isValid() || app.lastError().isValid())` just for cautiosness.

**Comment by paulolieuthier** [comment]
> Should we check the validity of QDBusInterface at all? (if the check is removed, in the above scenario nautilus is spawned correctly)

So the error is triggered but the app starts anyway? That's weird.

**Comment by palinek** [comment]
> > Should we check the validity of QDBusInterface at all? (if the check is removed, in the above scenario nautilus is spawned correctly)

> So the error is triggered but the app starts anyway? That's weird.

The error does not appear on the `call` itself but in the `QDBusInterface` constructor. After all I think we should not check the *validity* of the constructed object ...

**Comment by palinek** [comment]
> @palinek Do you have any clue on the origin of that error ? 
> I'm getting a QDBusError::InvalidMember in the second run. That's weird.
>
> It deserves more investigations, In the meantime we could only ignore QDBusError::InvalidMember
> Done here:
> https://github.com/luis-pereira/libqtxdg/tree/dbus_activate luis-pereira@741111b

I'm getting `QDBusError::Other`... so this probably isn't a good way.

**Comment by luis-pereira** [comment]
Probably, just probably  ;-)
We can dump the check, but we should have a reasoning IMO.

--
Luís Pereira
Sent from my Android
No dia 3 de Mar de 2016 06:53, "palinek" <notifications@github.com>
escreveu:

**Comment by palinek** [comment]
> We can dump the check, but we should have a reasoning IMO.

The reasoning here is that although the constructed interface is not valid (has lastError), the following `call` can/will succeed.

**Comment by paulolieuthier** [comment]
But it doesn't make sense to a DBusActivale app not to have its interface right. Or is it a problem with Qt?

**Comment by palinek** [comment]
> But it doesn't make sense to a DBusActivale app not to have its interface right. Or is it a problem with Qt?

I tend to say that the this `invalid` state of the interface is a *false positive* and is a result of some 
Qt caching....

Here, running the panel with QTDBUS_DEBUG=1:
- first start (no org.gnome.Nautilus service seen in qdbusviewer)
```
[debug] lxqt-panel (/home/palco/work/oss/lxqt/lxqt-panel/plugin-quicklaunch/quicklaunchaction.cpp:112) - execAction "/usr/share/applications/org.gnome.Nautilus.desktop"
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) sending message (blocking): QDBusMessage(type=MethodCall, service="org.freedesktop.DBus", path="/org/freedesktop/DBus", interface="org.freedesktop.DBus", member="GetNameOwner", signature="", contents=("org.gnome.Nautilus") )
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) sending message (blocking): QDBusMessage(type=MethodCall, service="org.gnome.Nautilus", path="/org/gnome/Nautilus", interface="org.freedesktop.DBus.Introspectable", member="Introspect", signature="", contents=() )
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) got message reply (blocking): QDBusMessage(type=MethodReturn, service=":1.79", signature="s", contents=("<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
                      "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
<!-- GDBus 2.46.2 -->
<node>
  <interface name="org.freedesktop.DBus.Properties">
    <method name="Get">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="s" name="property_name" direction="in"/>
      <arg type="v" name="value" direction="out"/>
    </method>
    <method name="GetAll">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="a{sv}" name="properties" direction="out"/>
    </method>
    <method name="Set">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="s" name="property_name" direction="in"/>
      <arg type="v" name="value" direction="in"/>
    </method>
    <signal name="PropertiesChanged">
      <arg type="s" name="interface_name"/>
      <arg type="a{sv}" name="changed_properties"/>
      <arg type="as" name="invalidated_properties"/>
    </signal>
  </interface>
  <interface name="org.freedesktop.DBus.Introspectable">
    <method name="Introspect">
      <arg type="s" name="xml_data" direction="out"/>
    </method>
  </interface>
  <interface name="org.freedesktop.DBus.Peer">
    <method name="Ping"/>
    <method name="GetMachineId">
      <arg type="s" name="machine_uuid" direction="out"/>
    </method>
  </interface>
  <interface name="org.gnome.Nautilus.FileOperations">
    <method name="CopyURIs">
      <arg type="as" name="SourceFilesURIList" direction="in"/>
      <arg type="s" name="DestinationDirectoryURI" direction="in"/>
    </method>
    <method name="EmptyTrash"/>
    <method name="CopyFile">
      <arg type="s" name="SourceFileURI" direction="in"/>
      <arg type="s" name="SourceDisplayName" direction="in"/>
      <arg type="s" name="DestinationDirectoryURI" direction="in"/>
      <arg type="s" name="DestinationDisplayName" direction="in"/>
    </method>
  </interface>
  <interface name="org.gtk.Actions">
    <method name="List">
      <arg type="as" name="list" direction="out">
      </arg>
    </method>
    <method name="Describe">
      <arg type="s" name="action_name" direction="in">
      </arg>
      <arg type="(bgav)" name="description" direction="out">
      </arg>
    </method>
    <method name="DescribeAll">
      <arg type="a{s(bgav)}" name="descriptions" direction="out">
      </arg>
    </method>
    <method name="Activate">
      <arg type="s" name="action_name" direction="in">
      </arg>
      <arg type="av" name="parameter" direction="in">
      </arg>
      <arg type="a{sv}" name="platform_data" direction="in">
      </arg>
    </method>
    <method name="SetState">
      <arg type="s" name="action_name" direction="in">
      </arg>
      <arg type="v" name="value" direction="in">
      </arg>
      <arg type="a{sv}" name="platform_data" direction="in">
      </arg>
    </method>
    <signal name="Changed">
      <arg type="as" name="removals">
      </arg>
      <arg type="a{sb}" name="enable_changes">
      </arg>
      <arg type="a{sv}" name="state_changes">
      </arg>
      <arg type="a{s(bgav)}" name="additions">
      </arg>
    </signal>
  </interface>
  <interface name="org.freedesktop.Application">
    <method name="Activate">
      <arg type="a{sv}" name="platform-data" direction="in">
      </arg>
    </method>
    <method name="Open">
      <arg type="as" name="uris" direction="in">
      </arg>
      <arg type="a{sv}" name="platform-data" direction="in">
      </arg>
    </method>
    <method name="ActivateAction">
      <arg type="s" name="action-name" direction="in">
      </arg>
      <arg type="av" name="parameter" direction="in">
      </arg>
      <arg type="a{sv}" name="platform-data" direction="in">
      </arg>
    </method>
  </interface>
  <interface name="org.gtk.Application">
    <method name="Activate">
      <arg type="a{sv}" name="platform-data" direction="in">
      </arg>
    </method>
    <method name="Open">
      <arg type="as" name="uris" direction="in">
      </arg>
      <arg type="s" name="hint" direction="in">
      </arg>
      <arg type="a{sv}" name="platform-data" direction="in">
      </arg>
    </method>
    <method name="CommandLine">
      <arg type="o" name="path" direction="in">
      </arg>
      <arg type="aay" name="arguments" direction="in">
      </arg>
      <arg type="a{sv}" name="platform-data" direction="in">
      </arg>
      <arg type="i" name="exit-status" direction="out">
      </arg>
    </method>
    <property type="b" name="Busy" access="read">
    </property>
  </interface>
  <node name="SearchProvider"/>
</node>
") )
[debug] lxqt-panel (unknown:0) - Adding rule: type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='org.gnome.Nautilus'
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) sending message (blocking): QDBusMessage(type=MethodCall, service="org.gnome.Nautilus", path="/org/gnome/Nautilus", interface="org.freedesktop.Application", member="Activate", signature="", contents=([QVariantMap]) )
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) got message reply (blocking): QDBusMessage(type=MethodReturn, service=":1.79", signature="", contents=() )
[debug] lxqt-panel (unknown:0) - Removing rule: type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='org.gnome.Nautilus'
```
- closing nautilus... and after some time
```
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x206b410) got message (signal): QDBusMessage(type=Signal, service=":1.1", path="/org/a11y/atspi/registry", interface="org.a11y.atspi.Registry", member="EventListenerDeregistered", signature="ss", contents=(":1.32", "") )
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x206b410) sending message (blocking): QDBusMessage(type=MethodCall, service="org.a11y.atspi.Registry", path="/org/a11y/atspi/registry", interface="org.a11y.atspi.Registry", member="GetRegisteredEvents", signature="", contents=() )
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x206b410) got message reply (blocking): QDBusMessage(type=MethodReturn, service=":1.1", signature="a(ss)", contents=([Argument: a(ss) {}]) )
```
- trying to start nautilus again (check is there, but continuing to `call` in all cases)
```
[debug] lxqt-panel (/home/palco/work/oss/lxqt/lxqt-panel/plugin-quicklaunch/quicklaunchaction.cpp:112) - execAction "/usr/share/applications/org.gnome.Nautilus.desktop"
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) sending message (blocking): QDBusMessage(type=MethodCall, service="org.freedesktop.DBus", path="/org/freedesktop/DBus", interface="org.freedesktop.DBus", member="GetNameOwner", signature="", contents=("org.gnome.Nautilus") )
[debug] lxqt-panel (unknown:0) - Adding rule: type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='org.gnome.Nautilus'
[warning] lxqt-panel (/home/palco/work/oss/lxqt/libqtxdg/xdgdesktopfile.cpp:467) - XdgDesktopFileData::startByDBus: can't call method on invalid interface: "Could not get owner of name 'org.gnome.Nautilus': no such name"
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) sending message (blocking): QDBusMessage(type=MethodCall, service="org.gnome.Nautilus", path="/org/gnome/Nautilus", interface="org.freedesktop.Application", member="Activate", signature="", contents=([QVariantMap]) )
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) got message reply (blocking): QDBusMessage(type=MethodReturn, service=":1.80", signature="", contents=() )
[debug] lxqt-panel (unknown:0) - Removing rule: type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='org.gnome.Nautilus'
[debug] lxqt-panel (unknown:0) - QDBusConnectionPrivate(0x1bc0f10) got message (signal): QDBusMessage(type=Signal, service="org.freedesktop.DBus", path="/org/freedesktop/DBus", interface="org.freedesktop.DBus", member="NameOwnerChanged", signature="sss", contents=("org.gnome.Nautilus", "", ":1.80") )
```

Now reading how the DBus activation [works](https://wiki.gnome.org/HowDoI/DBusApplicationLaunching#Install_a_D-Bus_service_file) (dbus daemon spawns the service/process upon request based on configuration).. IMO the false error state arises as following:
- 1. try to activate via dbus
  - panel calls (constructor of QDBusInterface) `destination=org.freedesktop.DBus path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=GetNameOwner
   string "org.gnome.Nautilus"` and gets response `org.freedesktop.DBus.Error.NameHasNoOwner`  => `QDBusInterface::lastError()` == `OtherError`
  - right after calls (still in QDBusInterface c-tor) `destination=org.gnome.Nautilus path=/org/gnome/Nautilus; interface=org.freedesktop.DBus.Introspectable; member=Introspect`
  - the dbus daemon spawns nautilus process, which in turn will respond to the `Introspect` request
  - => `QDBusInterface::lastError()` == `NoError`
- 2. try to activate via dbus (nautilus still alive)
  - panel calls (constructor of QDBusInterface) `destination=org.freedesktop.DBus path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=GetNameOwner
   string "org.gnome.Nautilus"` and gets the service owner => `QDBusInterface::lastError()` == `NoError`
  - panel doesn't call (constructor of QDBusInterface) the `Introspect` (as the previous result is probably cached) => `QDBusInterface::lastError()` == `NoError`
- 3. try to activate via dbus (nautilus already gone)
  - panel calls (constructor of QDBusInterface) `destination=org.freedesktop.DBus path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=GetNameOwner
   string "org.gnome.Nautilus"` and gets response `org.freedesktop.DBus.Error.NameHasNoOwner`  => `QDBusInterface::lastError()` == `OtherError`
  - panel doesn't call (constructor of QDBusInterface) the `Introspect` (as the previous result is probably cached) => `QDBusInterface::lastError()` == `OtherError`

**Comment by paulolieuthier** [comment]
OK. It's a Qt bug, then. GTM.

**Pull request merged**